File: Library/Classes/Core/Classes/UI/CSTableViewCell.m
//
//  Created by Rene Dohan on 10/24/12.
//

#import "CSTableViewCell.h"
#import "CSLang.h"

@implementation CSTableViewCell

- (void)layoutSubviews {
    super.layoutSubviews;
    runWith(self.onLayoutSubviews, self);
}

- (NSString *)reuseIdentifier {
    if (super.reuseIdentifier)return super.reuseIdentifier;
    return self.class.description;
}

@end

File: Renetik/Renetik/Classes/Core/Classes/View/CSTableViewCell.swift
//
// Created by Rene Dohan on 12/22/19.
//

import UIKit

public protocol CSTableHeightCalculatingCell {
    associatedtype Row
    func height(for data: Row) -> CGFloat
}

open class CSTableViewCell: UITableViewCell {

//    let layoutFunctions: CSEvent<Void> = event()

    @discardableResult
    public func layout(function: @escaping Func) -> Self {
        layoutFunctions.listen { function() }
        function()
        return self
    }

    @discardableResult
    public func layout<View: UIView>(_ view: View, function: @escaping (View) -> Void) -> View {
        layoutFunctions.listen { function(view) }
        function(view)
        return view
    }

    override open func layoutSubviews() {
        super.layoutSubviews()
        onLayoutSubviews()
//        updateLayout()
    }

//    private func onLayoutSubviews() {}

//    public func updateLayout() { layoutFunctions.fire() }

    override open var reuseIdentifier: String? {
        if super.reuseIdentifier.notNil { return super.reuseIdentifier }
        return type(of: self).description()
    }
}